### PR TITLE
Add MUIcache plugin

### LINKED ISSUE #88: MUIcache parser
The artefact can indicate executed GUI-based programs on a Windows system.

## References
* [https://forensafe.com/blogs/muicache.html](https://forensafe.com/blogs/muicache.html)
* [https://www.magnetforensics.com/blog/forensic-analysis-of-muicache-files-in-windows/](https://www.magnetforensics.com/blog/forensic-analysis-of-muicache-files-in-windows/)
* [https://youtu.be/ea2nvxN878s](https://youtu.be/ea2nvxN878s)

### PR DESCRIPTION
**Pull request by Zawadidone**
Closes #88

### CODE REVIEW

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                        for item in subkey.subkeys():
+                            yield from self._get_records(item)
+                else:
+                    yield from self._get_records(key)
+
+    def _get_records(self, key: RegistryKey) -> Generator[MuicacheRecord, None, None]:
+        for index, entry in enumerate(key.values()):
+            user = self.target.registry.get_user(key)
+            try:
+                if entry.name.endswith(tuple(self.FIELD_NAMES)):
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:76]:
```suggestion
                if entry.name.endswith(tuple(self.FIELD_NAMES)):
```
If you make `self.FIELD_NAMES` a tuple itself, you don't need to make it a tuple here.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from dissect.target.helpers.regutil import RegistryKey
+from dissect.target.plugin import Plugin, export
+
+MuicacheRecord = create_extended_descriptor([RegistryRecordDescriptorExtension, UserRecordDescriptorExtension])(
+    "windows/registry/muicache",
+    [
+        ("varint", "index"),
+        ("string", "name"),
+        ("string", "value"),
+        ("uri", "path"),
```
[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: The tests values are converted to strings and is checked against double-backslash Windows paths as seen here https://github.com/fox-it/flow.record/blob/main/tests/test_fieldtypes.py#L488.

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: Is it an issue that the fieldname `path` in this plugin is not detected as `windows_path`, but `posix_path`?

```bash
$ target-query -f muicache -t MSEDGEWIN10_20220708124036.tar -q --limit 3 | rdump -F path
[reading from stdin]
<windows/registry/muicache path=posix_path('LanguageList')>
<windows/registry/muicache path=posix_path('@C:\\Windows\\system32\\windows.storage.dll')>
<windows/registry/muicache path=posix_path('@C:\\Windows\\system32\\NetworkExplorer.dll')>

$ target-query -f muicache -t MSEDGEWIN10_20220708124036.tar -q --limit 3 | rdump -L -F path
[reading from stdin]
--[ RECORD 1 ]--
path = LanguageList
--[ RECORD 2 ]--
path = @C:\Windows\system32\windows.storage.dll
--[ RECORD 3 ]--
path = @C:\Windows\system32\NetworkExplorer.dll
```

[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/target/plugins/os/windows/regf/muicache.py]: If you don't cast it to a string that should work as intended:
```
path = _path.from_windows(path)
```

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: Okay that line is adjusted.

`rdump` works with csv, line and the default record output, but json(lines) results in errors. Should I create a issue in the flow.record project?

```bash
python query.py -f muicache -t MSEDGEWIN10_20220708124036.tar -q --limit 3 | rdump -J
[reading from stdin]
{"_type": "recorddescriptor", "_data": ["windows/registry/muicache", [["string", "hostname"], ["string", "domain"], ["varint", "index"], ["string", "name"], ["string", "value"], ["path", "path"], ["string", "regf_hive_path"], ["string", "regf_key_path"], ["string", "username"], ["string", "user_id"], ["string", "user_home"]]]}
Traceback (most recent call last):
  File "[...]/dissect.target/.tox/py3/bin/rdump", line 8, in <module>
    sys.exit(main())
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/utils.py", line 57, in wrapper
    return func(*args, **kwargs)
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/tools/rdump.py", line 212, in main
    record_writer.write(rec)
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/adapter/jsonfile.py", line 37, in write
    self._write(r)
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/adapter/jsonfile.py", line 33, in _write
    record_json = self.packer.pack(obj)
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/jsonpacker.py", line 93, in pack
    return json.dumps(obj, default=self.pack_obj, indent=self.indent)
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "[...]/dissect.target/.tox/py3/lib/python3.10/site-packages/flow/record/jsonpacker.py", line 71, in pack_obj
    raise Exception("Unpackable type " + str(type(obj)))
Exception: Unpackable type <class 'flow.record.fieldtypes.windows_path'>
```

[the remark this answers is not in the document]
Comment (yunzheng) [reply dissect/target/plugins/os/windows/regf/muicache.py]: Yes please

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: https://github.com/fox-it/flow.record/issues/29

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: Ah right

[the remark this answers is not in the document]
Comment (Zawadidone) [reply dissect/target/plugins/os/windows/regf/muicache.py]: https://github.com/fox-it/flow.record/issues/31

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
+from typing import Generator
+
+from flow.record.fieldtypes import path as _path
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:3]:
```suggestion
from flow.record.fieldtypes import path
```

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                            yield from self._get_records(item)
+                else:
+                    yield from self._get_records(key)
+
+    def _get_records(self, key: RegistryKey) -> Generator[MuiCacheRecord, None, None]:
+        for index, entry in enumerate(key.values()):
+            user = self.target.registry.get_user(key)
+            try:
+                if entry.name.endswith(self.FIELD_NAMES):
+                    path, name = entry.name.rsplit(".", 1)
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:72]:
```suggestion
                    entry_path, name = entry.name.rsplit(".", 1)
```

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                else:
+                    name = None
+                    path = entry.name.rsplit(",-", 1)[0]
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:75]:
```suggestion
                    entry_path = entry.name.rsplit(",-", 1)[0]
```

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+                value = entry.value
+                path = _path.from_windows(path)
+
+                # Filter on the type of the registry value
+                if type(value) is bytes:
+                    continue
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:82]:
```suggestion

                # Filter on the type of the registry value
                if isinstance(entry.value, bytes):
                    continue
```

File: dissect/target/plugins/os/windows/regf/muicache.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+                yield MuiCacheRecord(
+                    index=index,
+                    name=name,
+                    value=value,
+                    path=path,
```
Comment (Schamper) [inline dissect/target/plugins/os/windows/regf/muicache.py:88]:
```suggestion
                    value=entry.value,
                    path=path.from_windows(entry_path),
```

**Review by Schamper — CHANGES_REQUESTED** [review]
Small renaming changes.

**Comment by Schamper** [comment]
Is this waiting on one or more bug fixes in flow.record or is this good to merge?

**Comment by Zawadidone** [comment]
> Is this waiting on one or more bug fixes in flow.record or is this good to merge?

The bug is fixed in flow.record (https://github.com/fox-it/flow.record/issues/31) but dissect.target needs a dev version of the package (https://github.com/Zawadidone/dissect.target/blob/feature/add_plugin_muicache/setup.py#L63) to verify this. Can I adjust this or are dev versions not allowed?

`"flow.record>=3.0.dev,<4.0.dev"`

**Comment by Schamper** [comment]
Is the bug specific to this MR or could this be merged and the version of flow.record bumped in a later fix? You can always install the development version of flow.record locally until a new version of flow.record is released:

```
pip install --upgrade --pre flow.record
```
You'll get a warning about conflicting versions but you can safely ignore that.

**Comment by Zawadidone** [comment]
Okay merging is fine

**Review by Schamper — APPROVED** [review]

**Pull request merged**